### DISCUSSION TITLE
Null reference index

### DISCUSSION THREAD

**Discussion by DoctorMarvin21** [dotnet/csharplang#4851] [Language Ideas] [answered]
Return null if index is out of range:
`var a = array1[?0];`

If array has no element at index zero it will return null. It will work with reference types but could be expanded to value type if statement will return nullable:
`int[] array1 = { 1, 3, 4, 5, 6 };`
`int? a = array1[?0]; // "a" is 1`
`int? b = array1[?5]; // "b" is null`

Same could be done with ranges. If range exceeds array length it will return null:
`int[] array1 = { 1, 3, 4, 5, 6 };`
`int[] array2 = array1[?0..2]; // { 1, 3 }`
`int[] array3 = array1[?0..5]; // null`

This should help with situations when array is not null but index is out of range. With null-coalescing assignment this could be powerful tool for one line statements such as this one:
`A a = array1[?5] ?? new A();`

Using older syntax this could be achieved only with if-else statement.

**Comment by theunrepentantgeek** [comment]
I can't remember the last time I needed to put a guard statement around an array access to prevent out of bounds access and to provide a default value. 

In my experience, attempting to access an out of bounds value in an array is virtually certain evidence of a bug. The remedy is to fix the bug, not to mask the problem by patching around the array index.

For the vanishingly small number of situations where this is desirable behaviour, a trivial extension method does what you want:

``` csharp
public static T? Index<T>(this T[] arr, int index)
    => index < arr.Length ? arr[index] : default(T);
```

Note that it doesn't need to use an if-else statement ... and it will work with any array.

**Reply by Joe4evr** [reply]
Also, there's a good chance the JIT could eliminate the bounds check of the array access since it sees the index has been checked against the length right before.

**Comment by canton7** [comment, marked as answer]
There's already a method for this: [`Enumerable.ElementAtOrDefault`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.elementatordefault?view=net-5.0).


```cs
var b = array1.ElementAtOrDefault(10);
```

Yes you could argue it's a bit longer than `?`, but 1) It already exists, 2) It doesn't normally need to be used very often, and 3) It stands out as being something unusual.

**Reply by DoctorMarvin21** [reply]
Yes, I forgot about this new LINQ feature and can agree with all of 3 statements.

**Reply by canton7** [reply]
> new

It's been around since 2007!

**Reply by svick** [reply]
Keep in mind that `ElementAtOrDefault` does what it says: it gives you `default` if an element is not present, not `null`, which is different than what the proposed syntax does for value types.

Though you can work around that by using `.Cast<T?>` first.

**Reply by canton7** [reply]
That's true. Although, adding `.Cast<T?>()` means that `ElementAtOrDefault` becomes O(n), rather than being O(1) when used on collections which are indexable.

**Comment by TonyValenti** [comment]
I have a TryGet extension method for arrays that lets me treat them kind of like a dictionary.
